CNI: keep cni config on shutdown; taint instead, queue deletions

squeed opened the pull request:
This change is made up of several commits, which build on each other. It fixes a wart with how CNI, Containerd, and Dockershim interact, and prevents some possible outage scenarios.

Ultimately, I'd like to make much of this less necessary with an improved CNI STATUS verb, but that will take some time to roll out across the ecosystem.

### Asynchronous CNI delete
The CNI plugin now queues DEL for later submission, rather than failing. It tries to connect to the agent socket, and if that fails, it will write a file in `/run/cilium` with the pod's details. When the agent starts up, it will read that queue and process all pending deletes.

This fixes #22067, where an end-user found themselves with an unrecoverable cluster.  Cilium got descheduled but their node was at its pod limit. They couldn't start any pods, and they couldn't delete any pods without Cilium running. Not good.

### Taint the node when Cilium is shut down

On nodes where Cilium is scheduled, if it is not running, taint the node with NoSchedule. This will prevent new pods from being scheduled there. This has two goals:
1. Minimize ignorable errors from pods failing to start because Cilium can't handle a CNI ADD
2. Minimize pods being started with a non-Cilium network provider

### Preserve CNI configuration 

Lastly, we no longer delete the CNI configuration file when stopping the agent. This has several important effects:
- the node no longer goes NotReady, so it won't be removed from cloud LB backends just because cilium is being upgraded. This prevents unneeded churn.
- CNI DEL will still succeed, so pods can always be cleaned up
- No chance of pods being started with a different CNI provider

```release-note
The Cilium operator now taints nodes where Cilium is scheduled to run but is not running. 
This prevents pods from being scheduled on nodes without Cilium.
The CNI configuration file is no longer removed on agent shutdown. 
This means that pod deletion will always succeed; previously it would fail if Cilium was down for an upgrade.
This should help prevent nodes accidentally entering an unmanageable state.
It also means that nodes are not removed from cloud LoadBalancer backends during Cilium upgrades.
```


Fixes: #22067.

christarazi commented on pkg/lock/lockfile/lockfile_linux.go:
```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: Apache-2.0
```
We'll need to update the CODEOWNERS for this new file. Not sure what is appropriate? `cilium/sig-agent`?

squeed replied:
Do we? `pkg/lock` is already in CODEOWNERS, wouldn't that cover it?

christarazi replied:
Ah, yes. My eyes just glazed over the `pkg/lock` part and thought I saw `pkg/lockfile`. :+1:

christarazi commented on plugins/cilium-cni/main.go:
```diff
@@ -695,7 +688,7 @@ func cmdDel(args *skel.CmdArgs) error {
 	}
 
-	if conf.IpamMode == ipamOption.IPAMDelegatedPlugin {
+	if n.IPAM.Type != "" {
```
I'm not sure I understand why this change is necessary?

squeed replied:
`conf` is retrieved from the agent's API, but we need to be able to proceed even if its down. Good question.

christarazi replied:
That makes sense. But my main question was about the previous logic checking against `ipamOption.IPAMDelegatedPlugin` and now we aren't. How come?

christarazi replied:
@squeed ^

christarazi replied:
Any updates here?

squeed replied:
So, there are two separate configuration "knobs" here: whether or not there is a delegated IPAM plugin provided in the CNI configuration file, and whether or not the daemon is in delegated-ipam mode.

For CNI ADD, we check both: there needs to be user-added configuration in the IPAM section of the CNI configuration file, as well as the daemon configured for delegated ipam mode.

For DEL, we don't have the ability to query the daemon. So, we check only for a user-supplied IPAM plugin in the CNI configuration, and always use that.

This is actually safe. On the off chance that the user has supplied an IPAM plugin, but the daemon is not configured to use delegated IPAM, we will _still do the IPAM cleanup_ in the agent, by the very nature of the EndpointDelete API call. And the delegated IPAM plugin will get a DEL for which it doesn't have an ADD, which, according to the CNI spec, it must ignore.

squeed replied:
@christarazi make sense?

christarazi replied:
:+1:

christarazi commented on operator/watchers/node_taint_test.go:
```diff
@@ -420,6 +427,7 @@ func (n *NodeTaintSuite) TestNodeConditionIfCiliumAndNodeAreReady(c *check.C) {
 	mno = markNodeOptions{
 		RemoveNodeTaint:        true,
 		SetCiliumIsUpCondition: true,
+		SetNodeTaint:           false, // too annoying to test this with the same reactor
```
I'm not sure what this comment is trying to say

squeed replied:
Mostly explaining why this particular test isn't, well, testing everything. I'll clarify.

christarazi commented on Documentation/installation/taints.rst:
```diff
@@ -44,6 +44,9 @@ node. The mechanism works as follows:
 3. From this point on, pods will start being scheduled and running on the node,
    having their networking managed by Cilium.
 
+4. If Cilium is temporarily removed from the node, the operator will re-apply
```
Nit: `operator` -> `Operator`

christarazi reviewed (changes requested):
Looks mostly good to me, a few minor comments to address.

If I may share an opinion: I appreciate the PR description because it made reviewing the code much easier, however I'd hate to lose all that context because it isn't somewhere in the commit msg. What do you think of putting it somewhere in the commit(s)? The reason is because when bisecting, it's much easier to read all the context related to a change from the commit msg, rather than think "hmm, I wonder if there's more to this" and have to locate the PR behind it. 

What I've seen people do (and what I try to do as well) is paste the "important" commit msgs in the PR description as well, so people sort of have a cover letter to read before diving in, as you did in your PR, while also preserving the history in the commit(s) themselves.

On another note, for the release note: I'd suggest stripping the extra newlines as it's not going to format quite as nicely in the end. Something like this would be easier to read:

```
This is sentence 1 of the release note.
This is sentence 2 of the same release note.
And so on...
```

qmonnet commented on pkg/defaults/defaults.go:
```diff
@@ -106,6 +106,11 @@ const (
 	// PidFilePath is the path to the pid file for the agent.
 	PidFilePath = RuntimePath + "/cilium.pid"
 
+	// XXX doc
```
Please update (same for `DeleteQueueLockfile`).

The rest of the PR looks good to me, but I only took a very brief look at the Go code.

squeed replied:
ouch. good point.

qmonnet commented on pkg/lock/lockfile/lockfile_linux.go:
```diff
@@ -0,0 +1,118 @@
+	return &Lockfile{
+		fp: fp,
+	}, nil
+
```
Nit: blank line

sayboras reviewed (approved):
This PR is not only interesting but alsoo fun to read, love the comment in code :+1:.

LGTM :heavy_check_mark:

squeed committed:
operator: taint nodes where Cilium is scheduled but not running

This means that pods will not be scheduled where cilium is down. This
allows us to stop removing the CNI configuration file whenever we
restart the agent.

On nodes where Cilium is scheduled, if it is not running, taint the node with
NoSchedule. This will prevent new pods from being scheduled there. This has
two goals:

- Minimize ignorable errors from pods failing to start because Cilium can't
  handle a CNI ADD
- Minimize pods being started with a non-Cilium network provider

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
add simple Flock-based lockfile implementation

This adds a simple wrapper around Flock(2). It support read / write
locks, as well as non-blocking operation.

It is currently Linux only, since only Linux has support for
per-file-descriptor locking.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
plugin, daemon: support asynchronous deletion when the agent is down

It is considered bad practice for CNI deletions to fail. This can get
nodes in a fairly broken state, especially if somehow Cilium gets
descheduled -- then we might not be able to get Cilium back on the node!

The CNI plugin now queues DEL for later submission, rather than failing.
It tries to connect to the agent socket, and if that fails, it will write
a file in /run/cilium with the pod's details. When the agent starts up, it will
read that queue and process all pending deletes.

This fixes #22067, where an end-user found themselves with an unrecoverable
cluster. Cilium got descheduled but their node was at its pod limit. They
couldn't start any pods, and they couldn't delete any pods without Cilium
running. Not good.

The behavior is simple enough, but requires some subtlety around file
locking to make sure we don't miss a deletion.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
helm: enable node tainting when no running agent

This has cosmetic and usability benefits, since it mostly prevents pods
from being scheduled on nodes where they cannot actually be started. It
is not fool-proof, but it will prevent most ignorable errors about pods
failing to start. This commit enables the auto-taint feature.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
helm: set cni.uninstall default to false

Don't delete the CNI configuration file when stopping the agent. This has
several important effects:
- the node no longer goes NotReady, so it won't be removed from cloud LB
  backends just because cilium is being upgraded
- CNI DEL will still succeed, so pods can always be cleaned up
- No chance of pods being started with a different CNI provider

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

chancez commented on plugins/cilium-cni/lib/deletion_queue.go:
```diff
@@ -0,0 +1,133 @@
+// EndpointDelete deletes an endpoint given by an endpoint id, either
+// by directly accessing the API or dropping in a queued-deletion file.
+// endpoint-id is a qualified endpoint reference, e.g. "container-id:XXXXXXX"
+func (dc *DeletionFallbackClient) EndpointDelete(id string) error {
```
Should there be a cap on how many queued deletes we allow? Just in case it grows indefinitely?

squeed replied:
That's a good idea for a few reasons
- protect against some kind of kubelet total insanity
- prevent against a downgrade where the agent is no longer processing deletions

I'll add it.

squeed replied:
Done!

christarazi replied:
I agree this is a nice change. I wonder if it would also be useful if we had some sort of preemptive warning as a log msg as we approach the hard limit (maybe at 75%, 90%, 95%?). Given the CNI binary has no way to export metrics AFAIK, we don't really have visibility as to what's happening with this queue. Also open to other ideas that could give us visibility. Ideally, this is a metric and we ship an alert with it as well.

In general, one of the major goals for Cilium moving forward is better supportability and that includes visibility. One quick way to do that is to enumerate the hard limits in Cilium and ensure Cilium doesn't just all of a sudden stop to a grinding halt when it reaches them.

squeed replied:
That's a fair observation, and I also dislike systems that have surprises. In this case, it's difficult to export this as a metric because it's only in play when the agent is down :-).

The potential failure mode here - the queue filling up - is equivalent to the behavior before this PR: CNI deletion fails explicitly and is retried by the kubelet. Very different semantics than, say, bpf map exhaustion. Given that there is a clean fallback, I'd say it's OK not to add complicated guard-rails or alerting here.

christarazi replied:
Fair enough. A note in the code where we fallback would be good to call out as a comment so that future readers know that that was an explicit design decision. We can also revisit it in the future if this theoretical problem actually turns up in the wild and at least we will have documented our assumptions of it in the code.

qmonnet reviewed (approved):
Nice!

christarazi commented on daemon/cmd/deletion_queue.go:
```diff
@@ -0,0 +1,86 @@
+		return func() {}
+	}
+
+	log.Infof("Processing queued endpoint deletions")
```
Nit: no need for `Infof`

christarazi commented on daemon/cmd/deletion_queue.go:
```diff
@@ -0,0 +1,86 @@
+
+	lf, err := lockfile.NewLockfile(defaults.DeleteQueueLockfile)
+	if err != nil {
+		log.WithError(err).WithField("filename", defaults.DeleteQueueLockfile).Error("Failed to create ")
```
Nit: we already have a logfield for file paths, we should prefer this one for the sake of consistency especially in structured logging. Please also apply for each occurrence.

```suggestion
		log.WithError(err).WithField(logfields.Path, defaults.DeleteQueueLockfile).Error("Failed to create ")
```

christarazi replied:
This one feels more like a warning because Cilium is still proceeding anyway. 

```suggestion
		log.WithError(err).WithField("filename", defaults.DeleteQueueLockfile).Warn("Failed to create CNI delete directory queue lockfile. Proceeding with deletion anyway")
```

christarazi commented on daemon/cmd/deletion_queue.go:
```diff
@@ -0,0 +1,86 @@
+	defer cancel()
+	err = lf.Lock(ctx, true)
+	if err != nil {
+		log.WithError(err).Error("Failed to lock deletion directory, proceeding anyways")
```
```suggestion
		log.WithError(err).Warn("Failed to lock CNI deletion directory, proceeding anyways")
```

What are the consequences of this failure? What should the user know?

squeed replied:
I've been thinking about this. On the one hand, it means we might miss an endpoint deletion in the worst possible case. On the other hand, I really **really** don't want to ever block agent startup for something non-critical. Plus, we might as well proceed with cleanup regardless.

One possible error message would be "Failed to lock CNI deletion directory, proceeding anyways. This may cause CNI deletions to be missed" which sounds drastic but is technically correct.

Thoughts?

christarazi replied:
I think it's fair enough.

christarazi commented on daemon/cmd/deletion_queue.go:
```diff
@@ -0,0 +1,86 @@
+		log.WithError(err).Error("Failed to list deletion files")
+	}
+
+	log.Infof("processing %d queued deletion requests", len(files))
```
```suggestion
	files, err := filepath.Glob(defaults.DeleteQueueDir + "/*.delete")
	if err != nil {
		log.WithError(err).Error("Failed to list deletion files. Unable to process CNI deletion queue. Endpoint queued for deletion will not be deleted")
	} else { 
	  log.Infof("processing %d queued deletion requests", len(files))
	  }
```

christarazi commented on daemon/cmd/deletion_queue.go:
```diff
@@ -0,0 +1,86 @@
+		// get the container id
+		epID, err := os.ReadFile(file)
+		if err != nil {
+			log.WithError(err).WithField("filename", file).Error("failed to read deletion file")
```
```suggestion
			log.WithError(err).WithField("filename", file).Error("Failed to read deletion file; endpoint will not be deleted")
```

christarazi commented on daemon/cmd/deletion_queue.go:
```diff
@@ -0,0 +1,86 @@
+		}
+
+		if err := os.Remove(file); err != nil {
+			log.WithError(err).WithField("filename", file).Error("failed to remove processed deletion file")
```
```suggestion
			log.WithError(err).WithField("filename", file).Warn("failed to remove processed deletion file, however the endpoint was deleted")
```

christarazi reviewed (commented):
Apologies for extending this PR, but upon a second read, I noticed that the error logs are quite terse. I think we'd benefit from providing more context when there are warnings / errors, because we don't want the user to then think, "OK well I've got this error, but what should I do / what is Cilium going to do about it?".

Minor comments below.
